### PR TITLE
Add properties regarding azure spot instances

### PR DESCRIPTION
**Pull request by axbarsan**
Towards https://github.com/giantswarm/giantswarm/issues/15310

This PR adds configuration settings for Azure Spot Instances:
* Node pool creation:
	* Enabling or disabling the feature
	* Setting a maximum price for deallocating the node pool (or disabling deallocating based on pricing reasons)
* Node pool listing/searching:
	* Seeing if the feature is enabled or not
	* Seeing the maximum price for deallocating the node pool (or the lack of it)

### CODE REVIEW

File: spec/definitions.yaml [hunk 1]
```diff
@@ -1019,6 +1019,24 @@ definitions:
...
+            type: object
+            properties:
+              enabled:
+                description: |
+                  Whether the feature is enabled or not.
+                type: boolean
+                default: false
+              max_price:
+                description: |
+                  The maximum price that the node pool VM Size can reach before the node pool is deallocated.
```
Comment (whites11) [inline spec/definitions.yaml:1034]:
```suggestion
                  The maximum price that the single node pool VM instance can reach before it is deallocated.
```

File: spec/definitions.yaml [hunk 2]
```diff
@@ -1019,6 +1019,24 @@ definitions:
...
+                  By setting this value to `-1`, the price will be fixed to the value it had when the
```
Comment (whites11) [inline spec/definitions.yaml:1035]:
```suggestion
                  By setting this value to `-1`, the price will be fixed to the on-demand price.
```

File: spec/definitions.yaml [hunk 3]
```diff
@@ -1019,6 +1019,24 @@ definitions:
...
+                  node pool was created. The node pool will not be deallocated for pricing reasons.
```
Comment (whites11) [inline spec/definitions.yaml:1036]:
```suggestion
                  
```

**Review by whites11 — APPROVED** [review]

**Review by tuommaki — APPROVED** [review]

File: spec/definitions.yaml [hunk 4]
```diff
@@ -1019,6 +1019,24 @@ definitions:
...
+                type: boolean
+                default: false
+              max_price:
+                description: |
+                  The maximum price that a single node pool VM instance can reach before it is deallocated.
+                  By setting this value to `-1`, the maximum price will be fixed to the on-demand price of
+                  the instance.
+                  The value can be a floating point number with a maximum of 5 decimal places.
+                type: number
+                default: -1
```
Comment (marians) [inline spec/definitions.yaml:1039]: Using a negative one int as a default appears overly complicated and counter-intuitive to me.

I would propose to interpret both an absent `max_price` attribute and `max_price: 0` as the "use maximum" instead.

File: spec/definitions.yaml [hunk 5]
```diff
@@ -1034,6 +1052,20 @@ definitions:
             type: string
             description: |
               Azure virtual machine size used by all nodes in this pool.
+          spot_instances:
+            description: |
+              Attributes defining the presence and configuration of Spot Instances.
```
Comment (marians) [inline spec/definitions.yaml:1057]: I would expect a bit more information about what spot instances are and what this really means for the node pool. Ideally we would link out to a good Azure docs page for the first.

Comment (axbarsan) [reply spec/definitions.yaml:1057]: Shouldn't that information be in the docs, rather than the API spec?

Comment (marians) [reply spec/definitions.yaml:1057]: Yeah, also fine. Then a link to these docs would be great (once they exist).

File: spec/definitions.yaml [hunk 6]
```diff
@@ -1019,6 +1019,24 @@ definitions:
...
+            type: object
+            properties:
+              enabled:
+                description: |
+                  Whether the feature is enabled or not.
+                type: boolean
+                default: false
+              max_price:
+                description: |
+                  The maximum price that a single node pool VM instance can reach before it is deallocated.
```
Comment (marians) [inline spec/definitions.yaml:1034]: I suggest to add info on the currency for clarity

```suggestion
                  The maximum price in USD that a single node pool VM instance can reach before it is deallocated.
```

File: spec/definitions.yaml [hunk 7]
```diff
@@ -1019,6 +1019,24 @@ definitions:
...
+                  By setting this value to `0`, the maximum price will be fixed to the on-demand price of
+                  the instance.
+                  The value can be a floating point number with a maximum of 5 decimal places.
+                type: number
```
Comment (marians) [inline spec/definitions.yaml:1038]:
```suggestion
                type: number
                format: float
```

File: spec/definitions.yaml [hunk 8]
```diff
@@ -1034,6 +1053,20 @@ definitions:
...
+              Attributes defining the presence and configuration of Spot Instances.
+            type: object
+            properties:
+              enabled:
+                description: |
+                  Whether the feature is enabled or not.
+                type: boolean
+              max_price:
+                description: |
+                  The maximum price that a single node pool VM instance can reach before it is deallocated.
```
Comment (marians) [inline spec/definitions.yaml:1067]:
```suggestion
                  The maximum price in USD that a single node pool VM instance can reach before it is deallocated.
```

File: spec/definitions.yaml [hunk 9]
```diff
@@ -1034,6 +1053,20 @@ definitions:
...
+                  Find details on this attribute in the [addNodePool](#operation/addNodePool) operation.
+                type: number
```
Comment (marians) [inline spec/definitions.yaml:1069]:
```suggestion
                type: number
                format: float
```

File: spec/definitions.yaml [hunk 10]
```diff
@@ -1019,6 +1019,25 @@ definitions:
...
+                description: |
+                  Whether the feature is enabled or not.
+                type: boolean
+                default: false
+              max_price:
+                description: |
+                  The maximum price in USD that a single node pool VM instance can reach before it is deallocated.
+                  By setting this value to `0`, the maximum price will be fixed to the on-demand price of
+                  the instance.
+                  The value can be a floating point number with a maximum of 5 decimal places.
```
Comment (marians) [inline spec/definitions.yaml:1037]: Another question: what usage time unit is the price for? Minute/hour/day etc.

Comment (axbarsan) [reply spec/definitions.yaml:1037]: It's per hour

Comment (marians) [reply spec/definitions.yaml:1037]: Then please add the word "hourly" to the description

**Review by marians — APPROVED** [review]

**Pull request merged**